### PR TITLE
Feature/pebbledb

### PR DESCRIPTION
**Pull request by arnaucube**
- Abstract db tests
    - no new tests where created, where abstracted into `db/dbtest` (and updated to use `quicktest` for checks)
    - in badgerdb, the test that was in `db/badgerdb/batch_test.go` has been moved into `db/badgerdb/badger_test.go`
- Add [pebbledb](https://github.com/cockroachdb/pebble) db interface implementation

### CODE REVIEW

File: vochain/types.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ var (
 	ErrVoteDoesNotExist = fmt.Errorf("vote does not exist")
 	ErrProcessNotFound  = fmt.Errorf("process not found")
 	// keys; not constants because of []byte
-	headerKey    = []byte("header")
-	heightKey    = []byte("height")
+	headerKey = []byte("header")
+	// heightKey    = []byte("height") // unused
```
Comment (mvdan) [inline vochain/types.go]: remove it then?

File: db/badgerdb/badger_test.go [hunk 2]
```diff
@@ -1,124 +1,57 @@
...
 
 import (
-	"bytes"
-	"strconv"
+	"encoding/binary"
 	"testing"
 
+	qt "github.com/frankban/quicktest"
 	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/db/dbtest"
```
Comment (mvdan) [inline db/badgerdb/badger_test.go]: move this package to `db/internal/dbtest`. we don't want it to be importable by others.

File: db/badgerdb/badger_test.go [hunk 3]
```diff
@@ -1,124 +1,57 @@
...
-	if err := d.Iterate(nil, func(k, v []byte) bool {
-		noPrefixKeysFound++
-		return true
-	}); err != nil {
-		t.Fatal(err)
-	}
-	if noPrefixKeysFound != (prefix0NumKeys + prefix1NumKeys) {
-		t.Errorf("expected %d keys, found %d", prefix0NumKeys+prefix1NumKeys, noPrefixKeysFound)
-	}
+func TestBatch(t *testing.T) {
```
Comment (mvdan) [inline db/badgerdb/badger_test.go:26]: why not put this one in dbtest too?

File: db/pebbledb/pebledb.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+var _ db.ReadTx = (*ReadTx)(nil)
+
+// WriteTx implements the interface db.WriteTx
+type WriteTx struct {
+	batch *pebble.Batch
+}
+
+// check that WriteTx implements the db.ReadTx & db.WriteTx interfaces
+var _ db.ReadTx = (*WriteTx)(nil)
+var _ db.WriteTx = (*WriteTx)(nil)
```
Comment (mvdan) [inline db/pebbledb/pebledb.go:24]: writetx is already a readtx, so the first line here is redundant

File: db/pebbledb/pebledb.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// Get implements the db.WriteTx.Get interface method
+func (tx WriteTx) Get(k []byte) ([]byte, error) {
+	return ReadTx(tx).Get(k)
+}
+
+// Set implements the db.WriteTx.Set interface method
+func (tx WriteTx) Set(k, v []byte) error {
+	return tx.batch.Set(k[:], v, nil)
```
Comment (mvdan) [inline db/pebbledb/pebledb.go]: I think that `[:]` expression is a no-op, since `k` is already a slice.

File: db/pebbledb/pebledb.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+		LowerBound: prefix,
+		UpperBound: keyUpperBound(prefix),
+	}
+	iter := db.db.NewIter(iterOptions)
+	defer func() {
+		errC := iter.Close()
+		if err != nil {
+			return
+		}
+		err = errC
```
Comment (mvdan) [inline db/pebbledb/pebledb.go:146]: I don't think this error ends up anywhere, unless you use `) (err error) {` for the return parameter.

**Review by mvdan — APPROVED** [review]

File: db/pebbledb/pebledb.go (hunk 6 above)
Comment (arnaucube) [reply db/pebbledb/pebledb.go:146]: Right! Updated.

File: vochain/types.go (hunk 1 above)
Comment (arnaucube) [reply vochain/types.go]: Initially I did not remove it as was 'out of the scope' of this PR, but appeared in the goanalyze checks. But I'm agree with what you say, so I've remove it in the last update.

File: db/badgerdb/badger_test.go (hunk 3 above)
Comment (arnaucube) [reply db/badgerdb/badger_test.go:26]: This test checks that the badgerdb implementation gets an error of `badger.ErrTxnTooBig` (in the form of `db.ErrTxnTooBig`), so for the moment I thought to leave it as specific test of badgerdb package.

Comment (mvdan) [reply db/badgerdb/badger_test.go:26]: Hm, I see. I'll probably try something here in my follow-up PR. SGTM

File: db/badgerdb/badger.go [hunk 7]
```diff
@@ -27,7 +27,6 @@ type WriteTx struct {
 }
 
 // check that WriteTx implements the db.ReadTx & db.WriteTx interfaces
-var _ db.ReadTx = (*WriteTx)(nil)
```
Comment (ed255) [inline db/badgerdb/badger.go:30]: why?

File: db/pebbledb/pebledb.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
+package pebbledb
```
Comment (ed255) [inline db/pebbledb/pebledb.go:1]: the filename should be `pebbledb.go`

File: db/pebbledb/pebledb.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+// check that ReadTx implements the db.ReadTx interface
+var _ db.ReadTx = (*ReadTx)(nil)
+
+// WriteTx implements the interface db.WriteTx
+type WriteTx struct {
+	batch *pebble.Batch
+}
+
+// check that WriteTx implements the db.ReadTx & db.WriteTx interfaces
```
Comment (ed255) [inline db/pebbledb/pebledb.go:23]: missing check for interface ReadTx

File: db/badgerdb/badger.go (hunk 7 above)
Comment (arnaucube) [reply db/badgerdb/badger.go:30]: https://github.com/vocdoni/vocdoni-node/pull/301#discussion_r723009131

File: db/pebbledb/pebledb.go (hunk 9 above)
Comment (arnaucube) [reply db/pebbledb/pebledb.go:23]: https://github.com/vocdoni/vocdoni-node/pull/301#discussion_r723009131

File: db/pebbledb/pebledb.go (hunk 8 above)
Comment (arnaucube) [reply db/pebbledb/pebledb.go:1]: Right! I've missed a `b` :S

File: db/badgerdb/badger.go (hunk 7 above)
Comment (ed255) [reply db/badgerdb/badger.go:30]: Ok, got it from https://github.com/vocdoni/vocdoni-node/pull/301#discussion_r723009131

File: db/pebbledb/pebledb.go (hunk 9 above)
Comment (ed255) [reply db/pebbledb/pebledb.go:23]: Ok, got it from https://github.com/vocdoni/vocdoni-node/pull/301#discussion_r723009131

**Pull request merged**